pkg/apis/batch/validation/validation.go

```diff
@@ -485,9 +525,36 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
 }
 
 // ValidateJobStatusUpdate validates an update to a JobStatus and returns an ErrorList with any errors.
-func ValidateJobStatusUpdate(status, oldStatus batch.JobStatus) field.ErrorList {
+func ValidateJobStatusUpdate(job, oldJob *batch.Job) field.ErrorList {
 	allErrs := field.ErrorList{}
-	allErrs = append(allErrs, validateJobStatus(&status, field.NewPath("status"))...)
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(job, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// In particular, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete, batch.JobFailureTarget} {
+			if isConditionTrue(oldJob.Status.Conditions, cType) && !isConditionTrue(job.Status.Conditions, cType) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, fmt.Sprintf("disablement of %s=True condition", string(cType))))
+			}
+		}
+		// In order to improve performance we only validate indexes in case
+		// there is a change.
+		if job.Status.CompletedIndexes != oldJob.Status.CompletedIndexes {
+			if err := validateIndexesFormat(job.Status.CompletedIndexes, int(*job.Spec.Completions)); err != nil {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completedIndexes"), job.Status.CompletedIndexes, fmt.Sprintf("error parsing completedIndexes: %s", err.Error())))
+			}
+		}
+		if !ptr.Equal(job.Status.FailedIndexes, oldJob.Status.FailedIndexes) && job.Status.FailedIndexes != nil {
```

Comment:
```suggestion
		if job.Status.FailedIndexes != nil && !ptr.Equal(job.Status.FailedIndexes, oldJob.Status.FailedIndexes) {
```

Is it better to check for returning early if the `failedIndexes` isn't nil?

Reply:
yes, it is better, good call, done